src/perception/filter.hpp

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <numeric>
+#include <vector>
+
+/***
+ * A filter that combines multiple readings into one.
+ * A user defined proportion acts as a median filter that gets rids of outliers,
+ * which are then piped into a mean filter that averages out the values.
+ *
+ * @tparam T Reading type
+ */
+template<typename T>
+class Filter {
+private:
+    std::vector<T> mValues;
+    std::vector<T> mSortedValues;
+    // After sorting, what proportion in the middle values should we use.
+    double mProportion;
+    // How many readings we have.
+    // This will be capped at the capacity, since when we add when full we will overwrite the oldest value.
+    size_t mFilterCount = 0;
+    // Index to the current head.
+    // Note this is a circular buffer, so this will wrap around when we reach the end of the internal vector.
+    size_t mHead = 0;
+
+public:
+    Filter() : mValues(1), mProportion(0.0) {}
```

Comment:
Is the default constructor here ever going to be used?

Reply:
Yes - It's used implicitly in since `PersistentFiducial` has three `MeanMedianFilter` member variables.